linkerd/proxy/detect/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+use futures::{try_ready, Future, Poll};
+use linkerd2_error::Error;
+use linkerd2_io::{BoxedIo, Peek};
+use linkerd2_proxy_core as core;
+
+/// A strategy for detectin values out of a client transport.
+pub trait Detect<T> {
+    type Target;
+
+    /// If the target can be determined by the target alone (i.e. because it's
+    /// known to be a server-speaks-first target), Otherwise, the target is
+    /// returned as an error.
+    fn detect_before_peek(&self, target: T) -> Result<Self::Target, T>;
+
+    /// If the target could not be determined without peeking, then used the
+    /// peeked prefix to determine the protocol.
+    fn detect_peeked_prefix(&self, target: T, prefix: &[u8]) -> Self::Target;
+}
+
+#[derive(Debug, Clone)]
+pub struct Accept<D, A> {
+    detect: D,
+    accept: A,
+    peek_capacity: usize,
+}
+
+pub enum AcceptFuture<T, D: Detect<T>, A: core::listen::Accept<(D::Target, BoxedIo)>> {
```

Comment:
Nit: could these bounds maybe go in a `where` clause? this is a lot for one line...